backend/app/services/task_service.py

```diff
@@ -0,0 +1,134 @@
+from datetime import datetime, timedelta
+
+import pytz
+from fastapi import Depends, HTTPException, status
+from fastapi_pagination import paginate
+from fastapi_utilities import repeat_at
+from sqlalchemy import and_, create_engine
+from sqlalchemy.orm import Session, sessionmaker
+from starlette.background import BackgroundTasks
+
+from app.core.config import settings
+from app.core.constants import ADMIN, TASK_DELETED_MESSAGE, TASK_NOT_FOUND
+from app.core.mail import mail
+from app.db.crud import CRUDBase
+from app.db.database import get_db
+from app.logger import logger
+from app.models.task import Task
+from app.models.user import User
+from app.schema.task_schema import TaskCreateRequest, TaskUpdateRequest
+
+
+class TaskService(CRUDBase):
+    def __init__(
+        self,
+        db: Session = Depends(get_db),
+        background_tasks: BackgroundTasks = BackgroundTasks(),
+    ):
+        self.db = db
+        self.background_tasks = background_tasks
+        super().__init__(model=Task)
+
+    def create_task(self, user: dict, create_task_request: TaskCreateRequest):
+        create_task_request.user_id = user["id"]
+        return self.create(db=self.db, obj_in=create_task_request)
+
+    def get_all_tasks(
+        self, search_query, category, priority_level, due_date, status, user: dict
+    ):
+        filters = []
+
+        if user and user["role"] != ADMIN:
+            filters.append(Task.user_id == user["id"])
+
+        optional_filters = {
+            Task.title.ilike(f"%{search_query}%"): search_query,
+            Task.category == category: category,
+            Task.priority_level == priority_level: priority_level,
+            Task.due_date == datetime.fromisoformat(due_date)
+            if due_date
+            else None: due_date,
+            Task.status == status: status is not None,
+        }
+
+        for condition, value in optional_filters.items():
+            if value:
+                filters.append(condition)
+
+        tasks = self.db.query(Task).filter(and_(*filters))
+
+        return paginate(tasks.order_by(Task.created_at.desc()).all())
```

Comment:
`.all()` fetches all records before paginating, which can be inefficient for large datasets. Apply pagination and ordering before fetching records.